plane/src/drone/docker/mod.rs

```diff
@@ -187,6 +188,82 @@ impl PlaneDocker {
             .ok_or(anyhow::anyhow!("no stats found for {container_id}"))?
             .map_err(|e| anyhow::anyhow!("{e:?}"))
     }
+
+    /// Prune stopped backend containers that are older than the prune threshold.
+    /// Then, remove dangling images.
+    pub async fn prune(&self, since: DateTime<Utc>, auto_prune: bool) -> Result<()> {
+        tracing::info!("Pruning Docker containers and images.");
+
+        // Get a list of stopped containers with the label `dev.plane.backend`.
+        let candidates = self
+            .docker
+            .list_containers::<String>(Some(ListContainersOptions {
+                all: true,
+                filters: vec![
+                    ("label".to_string(), vec![PLANE_DOCKER_LABEL.to_string()]),
+                    ("status".to_string(), vec!["exited".to_string()]),
+                ]
+                .into_iter()
+                .collect(),
+                ..Default::default()
+            }))
+            .await?;
+
+        tracing::info!(
+            num_containers = candidates.len(),
+            "Removing stopped backend containers."
+        );
+        let mut skipped = 0;
+        for container in candidates {
+            let Some(container_created) = container.created else {
+                tracing::warn!("Received container without created timestamp.");
+                continue;
+            };
+
+            let LocalResult::Single(container_created) = Utc.timestamp_opt(container_created, 0)
+            else {
+                tracing::warn!(
+                    "Received container created timestamp that is not a valid UTC timestamp."
+                );
+                continue;
+            };
+
+            if container_created >= since {
+                tracing::info!(
+                    container_id = container.id.as_ref(),
+                    %since,
+                    %container_created,
+                    "Skipping container because it was created before the prune threshold."
+                );
+                skipped += 1;
+                continue;
+            }
+
+            let Some(container_id) = container.id else {
+                tracing::warn!("Received container without id.");
+                continue;
+            };
+            tracing::info!(container_id, "Removing container.");
+            if let Err(e) = self.docker.remove_container(&container_id, None).await {
+                tracing::error!(?e, container_id, "Error removing container.");
+            }
+        }
+
+        if skipped > 0 {
+            tracing::info!(skipped, "Skipped pruning of {} containers.", skipped);
+        }
+
+        if auto_prune {
+            match self.docker.prune_images::<&str>(None).await {
```

Comment:
Would `auto_prune: true` have us lose the image that we are periodically pulling and cost cold start times? I'm guessing this is a feature we'd want to leave off in Jamsocket?

Reply:
Yes, I think that's right on both counts.